pkcs11module.c

```diff
@@ -1322,30 +1547,462 @@ PHP_METHOD(Module, C_CreateObject) {
+        case CKA_REQUIRED_CMS_ATTRIBUTES:
+        case CKA_DEFAULT_CMS_ATTRIBUTES:
+        case CKA_SUPPORTED_CMS_ATTRIBUTES:
+        case CKA_ID:
+        case CKA_APPLICATION:
+        case CKA_URL:
+        case CKA_CHAR_SETS:
+        case CKA_ENCODING_METHODS:
+        case CKA_MIME_TYPES:
+        default: /* TBD */
+            return 0; /* any length */
+    }
+    return 0;
+}
 
-ZEND_BEGIN_ARG_INFO_EX(arginfo_C_FindObjects, 0, 0, 2)
+ZEND_BEGIN_ARG_INFO_EX(arginfo_C_FindObjectsInit, 0, 0, 2)
     ZEND_ARG_TYPE_INFO(0, session, IS_OBJECT, 0)
     ZEND_ARG_TYPE_INFO(0, template, IS_ARRAY, 0)
 ZEND_END_ARG_INFO()
 
+PHP_METHOD(Module, C_FindObjectsInit) {
+    CK_RV rv;
+
+    zval *session;
+    zval *template = NULL; /* PHP array */
+
+    CK_ULONG ulCount = 0; /* number of attributes in the search template */
+    CK_ATTRIBUTE_PTR pTemplate = NULL;
+
+    ZEND_PARSE_PARAMETERS_START(1, 2)
+        Z_PARAM_OBJECT_OF_CLASS(session, ce_Pkcs11_Session)
+        Z_PARAM_OPTIONAL
+        Z_PARAM_ARRAY(template)
+    ZEND_PARSE_PARAMETERS_END();
+
+    pkcs11_object *objval = Z_PKCS11_P(ZEND_THIS);
+    pkcs11_session_object *sessionobjval = Z_PKCS11_SESSION_P(session);
+
+    if (template) {
+        int i = 0; /* index of each template[i] */
+        zend_string *key;
+        zval *value;
+        zend_long h;
+
+        ulCount = zend_hash_num_elements(Z_ARRVAL_P(template));
+
+        pTemplate = (CK_ATTRIBUTE_PTR)ecalloc(sizeof(*pTemplate), ulCount);
+        ZEND_HASH_FOREACH_KEY_VAL(Z_ARRVAL_P(template), h, key, value) {
+            zend_string *keyi;
+            zval *valuei;
+            zend_long hi;
+
+            (void)key; /* unused */
+            (void)h; /* unused */
+            if (Z_TYPE_P(value) != IS_ARRAY) {
+                zend_throw_exception(zend_ce_exception, "Each element of template has to be an array", 0);
+                return ;
+            }
+
+            /* struct CK_ATTRIBUTE: [ int value['type'], string value['Value'] ] */
+            ZEND_HASH_FOREACH_KEY_VAL(Z_ARRVAL_P(value), hi, keyi, valuei) {
+
+                if (keyi && (strcmp(ZSTR_VAL(keyi), "type") == 0)) {
+                    if (Z_TYPE_P(valuei) != IS_LONG) {
+                        zend_throw_exception(zend_ce_exception, "'type' must be an integer", 0);
+                        return ;
+                    }
+                    pTemplate[i].type = (CK_ATTRIBUTE_TYPE)Z_LVAL_P(valuei);
+                    ZVAL_DEREF(valuei);
+                }
+                if (keyi && (strcmp(ZSTR_VAL(keyi), "Value") == 0)) {
+                    if (Z_TYPE_P(valuei) != IS_STRING) {
+                        zend_throw_exception(zend_ce_exception, "'type' must be a byte string", 0);
+                        return ;
+                    }
+                    pTemplate[i].pValue = Z_STRVAL_P(valuei);
+                    pTemplate[i].ulValueLen = (unsigned long)Z_STRLEN_P(valuei);
+                }
+            } ZEND_HASH_FOREACH_END();
```

Comment:
I know that we are adding the C_* methods on the Module object to stay true to the original API, but seeing as there are really no direct translation for the way templates are handled in C, I think that we should stick to a simple associative array here, as the Session::findObjects() uses and parses using parseTemplate()

Reply:
parseTemplate() is used with this update https://github.com/gamringer/php-pkcs11/pull/18/commits/5a035edcd11d20e6e704e7f260599df35f4083ce